Added functions for mpi support in qutensornet

Vinitha-balachandran opened the pull request.

alecandido commented:
Tests are missing

Vinitha-balachandran commented:
> Tests are missing

Added pytest. Need to install pytest-mpi.

alecandido commented:
> Added pytest. Need to install pytest-mpi.

`pytest-mpi` is a very little maintained project, with very minimal code.
https://github.com/aragilar/pytest-mpi/blob/master/src/pytest_mpi/__init__.py

The rather ideal situation in which it could be convenient to just vendor the part we're interested in.

What do you exactly need from `pytest-mpi`?

Vinitha-balachandran commented:
> > Added pytest. Need to install pytest-mpi.
> 
> `pytest-mpi` is a very little maintained project, with very minimal code. https://github.com/aragilar/pytest-mpi/blob/master/src/pytest_mpi/__init__.py
> 
> The rather ideal situation in which it could be convenient to just vendor the part we're interested in.
> 
> What do you exactly need from `pytest-mpi`?

To test mpi implementation. I thought you were referring that pytests are missing in your previous comment.

alecandido commented:
> I thought you were referring that pytests are missing in your previous comment.

Indeed

> To test mpi implementation.

I was looking for a much more specific answer. How are you planning to use the features of `pytest-mpi`?

Vinitha-balachandran commented:
> > I thought you were referring that pytests are missing in your previous comment.
> 
> Indeed
> 
> > To test mpi implementation.
> 
> I was looking for a much more specific answer. How are you planning to use the features of `pytest-mpi` in your test code?

I was looking for how pytests can be implemented while using mpi. @liweintu suggested this pytest-mpi which I am using now. It is in test_quimb_backend.py to test the function which parallelize the circuit evaluation. As commented in the file, pytest for mpi implementation can be checked by command mpirun -n 2 python -m pytest tests/test_quimb_backend.py --with-mpi. WIthout the --with-mpi , it will skip the mpi tests.

alecandido commented:
> As commented in the file, pytest for mpi implementation can be checked by command `mpirun -n 2 python -m pytest tests/test_quimb_backend.py --with-mpi`. WIthout the --with-mpi , it will skip the mpi tests.

This is a basic Pytest feature, no need of plugins to do that. You can define yourself the marker in the Pytest configs (not even needed to use executable code, just write in `pyproject.toml`, and Pytest will load it).

The official documentation about markers is here:
https://docs.pytest.org/en/8.1.x/how-to/mark.html
(it's pretty brief, and linking to a variety of examples with clarifications)

Vinitha-balachandran commented:
> > As commented in the file, pytest for mpi implementation can be checked by command `mpirun -n 2 python -m pytest tests/test_quimb_backend.py --with-mpi`. WIthout the --with-mpi , it will skip the mpi tests.
> 
> This is a basic Pytest feature, no need of plugins to do it. You can define yourself the marker in the Pytest configs (not even needed to use executable code, just write in `pyproject.toml`, and Pytest will load it).
> 
> The official documentation about markers is here: https://docs.pytest.org/en/8.1.x/how-to/mark.html (it's pretty brief, and linking to a variety of examples with clarifications)
Thanks for the suggestion. I have updated pyproject.toml to use for 2 cpus (just used minimum requirement) and tests are passing using the marker feature. Updated the files in the repo.
